crossvalidation.py

```diff
@@ -2,123 +2,188 @@
+        mesh = fenics.UnitSquareMesh(*params['mesh'])
+    else:
+        print(f"mesh must be an xml file name or (width, height), got {params['mesh']}")
+        sys.exit()
+    V = fenics.FunctionSpace(mesh, 'CG', 1)
+    W = pf.ProductFunctionSpace(V)
+
+    errs = []
+    if args.use_hyperopt:
+        l2_0, l2_1, smooth_0, smooth_1 = tuning_params
+        params['regularization']['l2'] = [l2_0, l2_1]
+        params['regularization']['smoothing'] = [smooth_0, smooth_1]
+
+    results = {'params': params}
+    for fold, (train, test) in enumerate(data.split(k=params["folds"], include_between=True)):
+        print(f"Doing fold {fold} with {params['method']}...")
+        print("\t".join(["", "total_loss", "error", "regularization", "smoothing"]))
+        results[fold] = {}
+        eqn = pf.HittingTimes(W, boundary, epsilon=params['boundary']['epsilon'])
+        control = eqn.control
+        # loss functionals
+        xy0, xy1 = train.pair_xy()
+        train_sd = pf.SpatialData(
+                train.genetic_data["divergence"].to_numpy(),
+                xy0, xy1,
+                W,
+        )
+        train_loss = pf.LossFunctional(train_sd, control, params['regularization'])
+        
+        xy0, xy1 = test.pair_xy()
+        test_sd = pf.SpatialData(
+                test.genetic_data["divergence"].to_numpy(),
+                xy0, xy1,
+                W,
+        )
+        test_loss = pf.LossFunctional(test_sd, control, params['regularization'])
+        
+        invp = pf.InverseProblem(eqn, train_loss)
+        m_hats, losses, optim_return = invp.optimize(
+                control, 
+                method=params['method'],
+                options=params['options'],
+        )
+        results[fold]['m_hats'] = m_hats
+        results[fold]['losses'] = losses
+        results[fold]['optim_return'] = optim_return
+        
+        # test set error
+        control.update(m_hats[-1])
+        u_hat = eqn.solve()
+        test_error = test_loss.l2_error(u_hat)
+        results[fold]['test_error'] = test_error
+        errs.append(test_error)
+        print(f"Done: test error {test_error}")
+
+
+    print(f'saving file {results_file}')
+    with open(results_file, 'wb') as file:
+        pickle.dump(results, file, protocol=pickle.HIGHEST_PROTOCOL)
     
-    # test set error
-    control.update(m_hats[-1])
-    u_hat = eqn.solve()
-    test_error = test_loss.l2_error(u_hat)
-    results[fold]['test_error'] = test_error
-    print(f"Done: test error {test_error}")
+    return np.mean(errs)
+
 
+if __name__ == "__main__":
 
-print(f'saving file {results_file}')
-with open(results_file, 'wb') as file:
-    pickle.dump(results, file, protocol=pickle.HIGHEST_PROTOCOL)
+    args = parse_args(sys.argv[1:])
+    
+    if args.use_hyperopt:
+        # define a search space
+        space = [hp.lognormal('l2_0', 0, 2), hp.lognormal('l2_1', 0, 2), hp.lognormal('smooth_0', 0, 2), hp.lognormal('smooth_1', 0, 2)] 
```

Comment:
we probably need these ranges to be specifiable somehow - either as options, or in the params json?

Reply:
agreed! I think arguments seem cleanest as they aren't needed for grid search, but the json files can be the same otherwise. Should there be a separate flag for each param, or should the two `l2_0` and `l2_1`  get the same values?

Reply:
I'll add an l2 and smooth flag for now, but let me know if we need a separate flag for each.